New fields to better support stockcollections

laceysanderson opened the pull request:
# Tripal 4 Core Dev Task

### Issue #2267

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
We need two new fields to best support the new Germplasm Collection content type. Specifically,

1. StockCollection Linker field to be added to all stock content types. This will list the collections a current stock is in and allow curators to add a stock to existing stock collections from the stock edit form.
2. DB linker field to be added to the stockcollection content type. This will indicate that a current stock collection is from a specific DB. We need this because stockcollection links to the db table through stockcollection_db but NOT to the dbxref table like all other instances in Chado. It looks like it was assumed that a Genebank is a single stockcollection and thus maps to a full db. It didn't take into account that some dbs might have multiple collections such as diversity panels or experiment-specific collections to indicate the exact germplasm an experiment measures.

## Tasks

- [x] StockCollection Linker Field
- [x] StockCollection Widget
- [x] StockCollection Formatter
- [x] Tests for StockCollection Field
- [x] Tests for StockCollection Widget
- [x] Add StockCollection field to all germplasm content types
- [x] DB Linker Field
- [x] DB Widget
- [x] DB Formatter
- [x] Tests for DB Field
- [x] Tests for DB Widget
- [x] Add DB field to germplasm collection content type

## Testing?
<!--- Please describe in detail how to test these changes. -->
<!--- Reviewers will use this section to test the submission! -->
<!--- If you've implemented PHPUnit tests, you can describe the test cases here. -->

### Manual Testing

**Tests for StockCollection Field**

1. Create a fresh docker and login.
2. Navigate to Tripal > Page Structure > Import type collection.
3. Select Germplasm Content Types (Chado) and click ‘Import’
4. Run the given ‘drush' command on the terminal and refresh the site.
<img width="1685" height="299" alt="Screenshot 2025-12-02 at 10 42 38 AM" src="https://github.com/user-attachments/assets/0319c68d-1756-4b10-89a4-58c42ff4db82" />

5. All the Germplasm Content Types should appear on the page.
6. Navigate to Tripal > Content > Add Tripal Content.
7. Click on Germpalsm Collection to create a stock collection.
<img width="1394" height="863" alt="Screenshot 2025-12-02 at 9 49 06 AM" src="https://github.com/user-attachments/assets/6570eebb-527b-4952-a8c8-444ab74adc02" />

8. Once it’s saved navigate again to Tripal > Content > Add Tripal Content.
9. Click on any of the stock content types.
10. The added widget must be shown on all the stock content type pages and they should list the created stock collections in the dropdown as shown below.
<img width="1004" height="233" alt="Screenshot 2025-12-02 at 10 07 12 AM" src="https://github.com/user-attachments/assets/8154406b-4574-48cb-8524-eff3fd39ef7a" />

11. Create a stock with the created collection linked.
<img width="1372" height="922" alt="Screenshot 2025-12-02 at 9 59 48 AM" src="https://github.com/user-attachments/assets/ca7a58de-9cdb-4458-8f00-4292583e093f" />

12. Once created the Selected collection should appear as shown below.
<img width="764" height="773" alt="Screenshot 2025-12-02 at 10 00 09 AM" src="https://github.com/user-attachments/assets/10b54e31-9eb2-4df6-8561-ca0b1b31f864" />

13. Try adding multiple collections and see if they are listed correctly.



**Tests for DB Linker field**

1. If you haven't done the manual testing for StockCollection Field above, follow steps 1-5 from above. 
2. Navigate to Tripal > Content > Add Tripal Content.
3. Click on Germpalsm Collection to create a stock collection.
4. The 'Database Name' field should appear as shown below.
<img width="1026" height="93" alt="Screenshot 2025-12-02 at 2 57 35 PM" src="https://github.com/user-attachments/assets/7a7baa81-edbc-47a5-b56e-76a9e9e2056b" />

5. Fill out the form to add a Germplasm Collection and select the database.
<img width="1431" height="1041" alt="Screenshot 2025-12-02 at 2 58 41 PM" src="https://github.com/user-attachments/assets/52f1d8bb-2cac-4d33-8eea-6d50c6b64bcd" />

6. After created, the formatter should contain the Database Name you linked as shown.
<img width="725" height="669" alt="Screenshot 2025-12-02 at 2 58 54 PM" src="https://github.com/user-attachments/assets/6c9050ee-ca7e-4e9e-94a1-ee9bf8916b79" />

laceysanderson committed:
Match PBO to values used in extension modules.

laceysanderson committed:
Make description a single value (cardinality 1).

laceysanderson committed:
Coding standard fixes to the stock linker field.

laceysanderson committed:
Bring back id because it is still used.

laceysanderson committed:
Start the stock collection field.

laceysanderson committed:
Missing config warning in test so remove default for now.

laceysanderson committed:
Add Stock collection formatter based on stock formatter.

laceysanderson committed:
Add stock collection widget based on stock widget.

laceysanderson committed:
Add crud test for stock collection field

laceysanderson committed:
Add stock collection field to germplasm content types.

laceysanderson committed:
Merge branch '4.x' into tv4g1-issue2267-stockcollectionFields

laceysanderson committed:
Add new terms to the test environment.

laceysanderson committed:
Set sequence for changed tables.

laceysanderson committed:
update term definitions.

laceysanderson committed:
fix scope issues

laceysanderson committed:
use new term ids

laceysanderson committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue2267-stockcollectionFields

vidu-wil committed:
Merge branch '4.x' into tv4g1-issue2267-stockcollectionFields

vidu-wil committed:
added the DB linker field to stockcollection

vidu-wil committed:
fixed an issue with some items not being listed, and changed the name

vidu-wil committed:
Merge branch '4.x' into tv4g1-issue2267-stockcollectionFields

vidu-wil committed:
changed the match limit and select limit to 20

vidu-wil committed:
Reverted the url change for Plant Breeding Ontology

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoStockCollectionTypeDefault.php:
```diff
@@ -0,0 +1,284 @@
+      $linker_pkey_col = $linker_schema_def['primary key'];
+      // The following should be the same as $base_pkey_col.
+      // @todo make sure it is.
+      $linker_left_col = array_keys($linker_schema_def['foreign keys'][$base_table]['columns'])[0];
```
See PR #2381 for how this line should be changed in all fields.

laceysanderson reviewed (commented):
You are going to want to merge in #2389, add a generateSampleValue() implementation for both field types and add a scenario or two to tripal_chado/tests/src/Kernel/Plugin/ChadoField/FieldStaticMethodTest-TestInfo.yml for each field in order to get full coverage of your field type classes (except for isCompatible and discover.

vidu-wil committed:
Merge branch '4.x' into tv4g1-issue2267-stockcollectionFields

vidu-wil committed:
updated two field type classes with generateSampleValues() method and added testing for linker

vidu-wil committed:
added comments regarding no extra linker columns on the linker tables

dsenalik commented on tripal_chado/src/Plugin/Field/FieldFormatter/ChadoStockCollectionFormatterDefault.php:
```diff
@@ -0,0 +1,86 @@
+   */
+  public static function defaultSettings() {
+    $settings = parent::defaultSettings();
+    $settings['token_string'] = '[name] ([uniquename])';
```
So the default title format for the germplasmcollection content type is `[germcollection_unique_name][: [germcollection_name]]` - the two parts are in the opposite order as here. Personally I don't like that format as default, I would just prefer `[germcollection_name]`, but that is unrelated to this PR. Plus name can be null! Given that as the default, I think the default here should be the same, or change the germplasmcollection format if you prefer, as long as they are consistent.
```suggestion
    $settings['token_string'] = '[uniquename]: [name]';
```

dsenalik committed:
Merge branch '4.x' into tv4g1-issue2267-stockcollectionFields

dsenalik committed:
support custom chado tables

dsenalik commented on tripal_chado/config/install/tripal.tripalfield_collection.germplasm_chado.yml:
```diff
@@ -988,6 +1072,33 @@ fields:
+        content_type: 'germcollection'
+        label: 'Database Name'
+        type: 'chado_stockcollection_db_type_default'
+        description: 'Indicates that a current stock collection is from a specific DB.'
```
```suggestion
        description: 'Indicates that a stock collection is from a specific database.'
```
I feel like "current" is a bit confusing and not needed.

dsenalik commented on tripal_chado/src/Plugin/Field/FieldFormatter/ChadoStockCollectionFormatterDefault.php:
```diff
@@ -0,0 +1,86 @@
+#[TripalFieldFormatter(
+  id: 'chado_stockcollection_formatter_default',
+  label: new TranslatableMarkup('Chado Stock Collection formatter'),
+  description: new TranslatableMarkup('A simple list of the collections the current stock is part of.'),
```
```suggestion
  description: new TranslatableMarkup('A simple list of the collections the stock is part of.'),
```
Again don't think "current" adds any info.

dsenalik commented on tripal_chado/src/Plugin/Field/FieldFormatter/ChadoStockCollectionDbFormatterDefault.php:
```diff
@@ -0,0 +1,79 @@
+#[TripalFieldFormatter(
+  id: 'chado_stockcollection_db_formatter_default',
+  label: new TranslatableMarkup('Chado Stock Collection DB Formatter'),
+  description: new TranslatableMarkup('Displays the specific database.'),
```
```suggestion
  description: new TranslatableMarkup('Displays the specific database that a stock collection links to.'),
```

dsenalik commented on tripal_chado/src/Plugin/Field/FieldType/ChadoStockCollectionDbTypeDefault.php:
```diff
@@ -0,0 +1,276 @@
+  id: 'chado_stockcollection_db_type_default',
+  category: 'tripal_chado',
+  label: new TranslatableMarkup('Chado Stock Collection DB reference'),
+  description: new TranslatableMarkup('Indicates that a current stock collection is from a specific DB.'),
```
```suggestion
  description: new TranslatableMarkup('Indicates that a stock collection is from a specific database.'),
```

dsenalik commented on tripal_chado/src/Plugin/Field/FieldType/ChadoStockCollectionTypeDefault.php:
```diff
@@ -0,0 +1,299 @@
+  id: 'chado_stockcollection_type_default',
+  category: 'tripal_chado',
+  label: new TranslatableMarkup('Chado Stock Collection'),
+  description: new TranslatableMarkup('Indicate that the current chado stock is part of the indicated stock collection.'),
```
```suggestion
  description: new TranslatableMarkup('Indicate that a chado stock is part of the indicated stock collection.'),
```

dsenalik committed:
rename file+class for consistency and coding standards

dsenalik committed:
rename file and class for consistency

dsenalik committed:
rename file and class for consistency

dsenalik reviewed (changes requested):
So far looks great!
 - A couple of wording suggestions.
 - Commit https://github.com/tripal/tripal/pull/2268/commits/eac8ac9f3d976f7b75fbd555d17caf60fceb6791 allows us to support chado custom tables
 - A title format suggestion.
 - I pushed some commits renaming files to be consistent between type, formatter, and widget https://github.com/tripal/tripal/pull/2268/commits/9a221a985f5050a9c2fb1fcee5f9977c62816061 https://github.com/tripal/tripal/pull/2268/commits/f37a088b06ce9a9f7a23dc5074675b31cf01d75a https://github.com/tripal/tripal/pull/2268/commits/52ad5d07776690f277b9b12b80a8b6ef52b41f43
 - So the list of collections has `[collection of specimens]` appended to each one, I don't think this should be included because it is the content type term and will always be the same for all of them.
<img width="563" height="273" alt="PR2268 1" src="https://github.com/user-attachments/assets/3bf56828-cbeb-4bdd-8ce5-845535c7b817" />

 - One more significant change to make. First I made a database
`INSERT INTO db (name, description, urlprefix, url) VALUES ('GRIN Accession', 'The Germplasm Resources Information Network (GRIN) web server provides germplasm information about plants, animals, microbes and invertebrates. This program is within the U.S. Department of Agriculture''s Agricultural Research Service', 'https://npgsweb.ars-grin.gov/gringlobal/accessiondetail.aspx?accid={accession}', 'https://www.ars-grin.gov/');`
If I link a stock collection to this database, it is displayed like this:
<img width="178" height="82" alt="PR2268 2" src="https://github.com/user-attachments/assets/ef5029f0-8ae0-4753-bd8a-9423c491c280" />

but the URL (when present, equal to `https://www.ars-grin.gov/` here) should cause it to be clickable. We need another property in the Type and handler for that in the Formatter.

dsenalik committed:
run tests in separate process

vidu-wil committed:
Applied the suggested change.

Co-authored-by: Douglas Senalik <dsenalik@wisc.edu>

vidu-wil committed:
Applied the suggested change

Co-authored-by: Douglas Senalik <dsenalik@wisc.edu>

vidu-wil committed:
Applied the suggested change

Co-authored-by: Douglas Senalik <dsenalik@wisc.edu>

vidu-wil committed:
Applied the suggested change

Co-authored-by: Douglas Senalik <dsenalik@wisc.edu>

vidu-wil committed:
Applied the suggested change

Co-authored-by: Douglas Senalik <dsenalik@wisc.edu>

vidu-wil committed:
Applied the suggested change

Co-authored-by: Douglas Senalik <dsenalik@wisc.edu>

vidu-wil committed:
Merge branch '4.x' into tv4g1-issue2267-stockcollectionFields

dsenalik commented:
Here's another thought, the name in the select list
<img width="563" height="273" alt="PR2268 1" src="https://github.com/user-attachments/assets/3bf56828-cbeb-4bdd-8ce5-845535c7b817" />
is the "name" field, which chado allows to be NULL.
I think here too we should match the title format and have 
uniquename: name
that way if name is null you can still select the collection.

reynoldtan commented:
I conducted the manual test and verified that the expected out comes were achieved.

- [x] Type collection import completed without errors.
<img width="379" height="232" alt="Screen Shot 2026-02-06 at 11 50 39 AM" src="https://github.com/user-attachments/assets/c318c6c9-6172-42cb-936b-2e11629d1a28" />

- [x] Germplasm Content Types appear on page.
<img width="516" height="282" alt="Screen Shot 2026-02-06 at 11 52 35 AM" src="https://github.com/user-attachments/assets/1108833d-153e-4874-8524-db5ad9d25190" />

- [x] A widget for collection is present in the form.
<img width="460" height="197" alt="Screen Shot 2026-02-06 at 11 53 20 AM" src="https://github.com/user-attachments/assets/a6013c41-9324-4a6f-8249-1059cc831a2f" />

- [x] Content and collection are linked correctly.
<img width="315" height="585" alt="Screen Shot 2026-02-06 at 11 57 04 AM" src="https://github.com/user-attachments/assets/2c75669c-754b-4bf0-9c32-3589d7f10925" />

- [x] Collections are listed correctly.
<img width="460" height="197" alt="Screen Shot 2026-02-06 at 11 53 20 AM" src="https://github.com/user-attachments/assets/267b2aa9-1c48-4949-a2d0-ed92cbb39701" />

- [x] Database Linker field is referenced by the formatter.
<img width="353" height="522" alt="Screen Shot 2026-02-06 at 12 00 06 PM" src="https://github.com/user-attachments/assets/318c2413-3349-4148-8a99-07f15c437bcc" />

laceysanderson committed:
Merge branch '4.x' into tv4g1-issue2267-stockcollectionFields

laceysanderson committed:
Add url to the field type and use it to make a link where there is not an entity.

laceysanderson committed:
Missed updating the tests.

laceysanderson committed:
Fix group to match standards.

vidu-wil committed:
removed the type from select options

laceysanderson commented:
## Remove type from drop-down

> So the list of collections has [collection of specimens] appended to each one, I don't think this should be included because it is the content type term and will always be the same for all of them.

Addressed in [b5b0eb3](https://github.com/tripal/tripal/pull/2268/commits/b5b0eb3e7a4ea2f6f570972cbfda37aa21045d21).

<img width="1073" height="768" alt="Screenshot 2026-02-09 at 5 26 15 PM" src="https://github.com/user-attachments/assets/4f21d6b3-88d2-44b6-b483-5603de857c71" />

## Make DB link if URL provided.

> One more significant change to make. First I made a database
> ```
> INSERT INTO db (name, description, urlprefix, url) VALUES ('GRIN Accession', 'The Germplasm Resources Information Network (GRIN) web server provides germplasm information about plants, animals, microbes and invertebrates. This program is within the U.S. Department of Agriculture''s Agricultural Research Service', 'https://npgsweb.ars-grin.gov/gringlobal/accessiondetail.aspx?accid={accession}', 'https://www.ars-grin.gov/');
> ```
> If I link a stock collection to this database, it is displayed like this but the URL (when present, equal to https://www.ars-grin.gov/ here) should cause it to be clickable. We need another property in the Type and handler for that in the Formatter.

Addressed in [d084b98](https://github.com/tripal/tripal/pull/2268/commits/d084b98f5d1e90af903db57fed60f01b9950331c).

<img width="859" height="504" alt="Screenshot 2026-02-09 at 5 27 39 PM" src="https://github.com/user-attachments/assets/a92544d8-d06e-426a-9134-cb4f2856271e" />

## Format drop-down with uniquename + name

> Here's another thought, the name in the select list is the "name" field, which chado allows to be NULL. I think here too we should match the title format and have `uniquename: name` that way if name is null you can still select the collection.

The drop-down is populated using the generic autocomplete controller with the column name set to `name`. However, this controller only supports using a single column for this. To accomplish what you requested @dsenalik we either need to create a custom controller or add this functionality to the generic controller. Should we do that in this PR? or save it for a later one?

dsenalik commented:
> To accomplish what you requested @dsenalik we either need to create a custom controller or add this functionality to the generic controller. Should we do that in this PR? or save it for a later one?

I didn't think of this! This is not a high priority, so I think I will create a separate issue to implement this.

Issue #2422

laceysanderson commented:
Awesome, thanks @dsenalik! This is ready for another review then as I think we addressed all your comments 😁

dsenalik reviewed (approved):
I redid testing and have no more complaints!
Much appreciated as I will be using the stock collection type 😍
I did test collections with NULL names - they actually work correctly, even though you can't see a name 🙈 you can select them.

<img width="727" height="223" alt="Image" src="https://github.com/user-attachments/assets/434b1523-9be5-4a83-8a4f-50c1979c578c" />

We can fix this in #2422 someday.
